File: src/main/java/ting/leetcode/java/linkedlist/ReverseLinkedList.java
package ting.leetcode.java.linkedlist;

public class ReverseLinkedList {
    public ListNode reverseList(ListNode current) {
        if (current == null || current.next == null) {
            return current;
        }

        // 透過遞迴找到最後一個節點，也就是反轉後的頭部節點
        ListNode last = reverseList(current.next);
        current.next.next = current;
        current.next = null;
        return last;
    }

    // 另一種反轉法，結就較為簡單
    public ListNode reverse(ListNode head, ListNode prev) {
        if (head == null) return prev;
        ListNode newHead = head.next;
        head.next = prev;
        return reverse(newHead, head);
    }

    public ListNode reverseListByLoop(ListNode current) {
        if (current == null) {
            return null;
        }

        ListNode prev = null;
        while (current != null) {
            ListNode next = current.next;
            current.next = prev;
            prev = current;
            current = next;
        }
        return prev;
    }
}

File: src/main/kotlin/ting/leetcode/linkedlist/ReverseLinkedList.kt
package ting.leetcode.linkedlist

class ReverseLinkedList {
    fun reverseList(head: ListNode?): ListNode? {
        if (head?.next == null) return head
        val nodes = mutableListOf<ListNode?>()
        var currentNode = head
        while (currentNode != null && nodes.add(currentNode))
            currentNode = currentNode.next.also { currentNode?.next = null }

        nodes.reverse()
        nodes.forEachIndexed { index, node ->
            if (index + 1 <= nodes.lastIndex) node?.next = nodes[index + 1]
        }

        return nodes.first()
    }

    // 反轉 link 方向，第一個 link 指向 null
    fun reverseListByReversingLink(head: ListNode?): ListNode? {
        var currentNode = head
        var reverseNode: ListNode? = null

        while (currentNode != null) {
            reverseNode = ListNode(currentNode.`val`).also { it.next = reverseNode }
            currentNode = currentNode.next
        }

        return reverseNode
    }
}
